File: python/001.strategy.py
class SortStrategy:
    def sort(arr):
        pass


class BubbleSortStrategy(SortStrategy):
    def sort(self, arr):
        print('do buble sort')
        return arr


class QuickSortStrategy(SortStrategy):
    def sort(self, arr):
        print('do quick sort')
        return arr


class Sorter:
    def __init__(self, sortStrategy):
        self.sortStrategy = sortStrategy

    def sort(self, arr):
        return self.sortStrategy.sort(arr)


Sorter(QuickSortStrategy()).sort([1, 5, 3, 2, 4])

File: typescript/001.strategy.js
"use strict";
var BubbleSortStrategy = /** @class */ (function () {
    function BubbleSortStrategy() {
    }
    BubbleSortStrategy.prototype.sort = function (arr) {
        console.log('do bubble sort');
        return arr;
    };
    return BubbleSortStrategy;
}());
var QuickSortStrategy = /** @class */ (function () {
    function QuickSortStrategy() {
    }
    QuickSortStrategy.prototype.sort = function (arr) {
        console.log('do quick sort');
        return arr;
    };
    return QuickSortStrategy;
}());
var Sorter = /** @class */ (function () {
    function Sorter(sortStrategy) {
        this.sortStrategy = sortStrategy;
    }
    Sorter.prototype.sort = function (arr) {
        return this.sortStrategy.sort(arr);
    };
    return Sorter;
}());
new Sorter(new BubbleSortStrategy()).sort([9, 2, 4, 3, 5]);
new Sorter(new QuickSortStrategy()).sort([9, 2, 4, 3, 5]);
